%3a1/srfi-1-reference.scm

```diff
@@ -733,138 +773,192 @@
 ;;; However, if any of the lists is empty, just abort and return [() ()].
 
 (define (%cars+cdrs lists)
-  (let f ((ls lists) (a* '()) (d* '()))
+  (let f ((ls lists))
```

Comment:
What's the real performance difference between doing these with or without tail-recursion?

Reply:
Ok, I'll see if I can run some performance tests.

Reply:
These are both asymptotically the same, so I suspect this will be a test of whether the stack call infrastructure on the benchmarked Scheme implementation is fast enough to keep up with the extra set of CONS operations with the extra GC overhead. I suspect that it will be better, but I'm curious to know what sort of difference this really creates.

Reply:
Some preliminary tests with larceny list benchmarks show not much difference in time (it seems random), and a little difference in allocation (about 6 or 7% less in new implementation).
For old lists implementation:
```
> (time (go 10))
(time (go 10))
110 collections
71.538586836s elapsed cpu time, including 1.563913807s collecting
71.540591993s elapsed real time, including 1.564609554s collecting
939307584 bytes allocated, including 492455520 bytes reclaimed

```
For new lists implementation:
```
> (time (go 10))
(time (go 10))
104 collections
73.248517936s elapsed cpu time, including 1.534103670s collecting
73.263589521s elapsed real time, including 1.535928427s collecting
883110336 bytes allocated, including 452413824 bytes reclaimed

```

Benchmark code https://github.com/larcenists/larceny/blob/master/test/Benchmarking/R7RS/src/list.scm

Reply:
Well, you might not believe this, but I've implemented a procedure that computes all possible sums of elements of a random set of integers and, well, the new code runs 2000 times faster for a list of length 17. I tested with lists of several lengths, the difference in speed is always very big.

I honestly don't know what happened here LOL I have no idea why all this difference.

```
;; Tests with new implementation
> (import (srfi :1))
> (load "list-bench.scm")
> (go 4 100)
((0 ()) (8 (8)) (49 (49)) (57 (8 49)) (64 (64)) (72 (8 64))
(95 (95)) (103 (8 95)) (113 (49 64)) (121 (8 49 64))
(144 (49 95)) (152 (8 49 95)) (159 (64 95)) (167 (8 64 95))
(208 (49 64 95)) (216 (8 49 64 95)))
;; The code seems to work
> (time (take (go 17 10000) 3))
(time (take (go 17 ...) ...))
2 collections
0.037925419s elapsed cpu time, including 0.013374109s collecting
0.037921060s elapsed real time, including 0.013384622s collecting
12589776 bytes allocated, including 15729248 bytes reclaimed
((0 ()) (273 (273)) (869 (869)))
> 
It's been nice interacting with you!
Press C-c C-z to bring me back.
Chez Scheme Version 9.5.1
Copyright 1984-2017 Cisco Systems, Inc.

;; Tests with old implementation
> (import (srfi :1 old lists))
> (load "list-bench.scm")
> (go 4 100)
((0 ()) (7 (7)) (19 (19)) (26 (7 19)) (92 (92)) (98 (98))
(99 (7 92)) (105 (7 98)) (111 (19 92)) (117 (19 98))
(118 (7 19 92)) (124 (7 19 98)) (190 (92 98))
(197 (7 92 98)) (209 (19 92 98)) (216 (7 19 92 98)))
;; The code seems to work.
> (time (take (go 17 10000) 3))
(time (take (go 17 ...) ...))
2 collections
78.177356480s elapsed cpu time, including 0.016559823s collecting
78.175950900s elapsed real time, including 0.016571037s collecting
17833008 bytes allocated, including 8630096 bytes reclaimed
((0 ()) (408 (408)) (1279 (1279)))
```

Reply:
Source code of tests run:
[](https://paste.ofcode.org/4YpFnUS2GVDj5cz7kqypEx)

Reply:
Works for me. :-)